README.md
# Loop server

[![Build Status](https://travis-ci.org/mozilla-services/loop-server.svg?branch=master)](https://travis-ci.org/mozilla-services/loop-server)

This is the server part of the Loop project. You can find more information on
its APIs by looking at [the online documentation](https://docs.services.mozilla.com/loop/)

## How do I install loop server?

You will need [Redis](http://redis.io/) server installed:

### Linux

```sh
apt-get install redis-server
```

or

```sh
yum install redis
```

### OS X

Assuming you have [brew](http://brew.sh/) installed, use it to install Redis:

```sh
brew install redis
```

If you need to restart it (after configuration update):

```sh
brew services restart redis
```

### All Platforms

Then clone the Loop server and install its dependencies:

```sh
git clone https://github.com/mozilla/loop-server.git
cd loop-server && make install
```

## How do I run it?

You can create your configuration file in `config/{NODE_ENV}.json`

`dev` is the environment by default. In order to run the server, you'll need to
create a `dev.json` file in the config folder. You can do so by using this
command:

```sh
cp config/{sample,dev}.json
```

Be sure to edit the content of `config/dev.json`. You'll especially need to
specify your [TokBox](http://tokbox.com/) credentials.

Once that's done, you can do:

```sh
make runserver
```

(which is equivalent to `NODE_ENV=dev make runserver`)

## How do I run the tests?

```sh
make test
```

Redis is the default backend. The code is made to support multiple ones but
only supports Redis for now.

In order to have tests working with Mac OS, make sure your `ulimit`
value is high enough or you will get EMFILE errors:

```sh
ulimit -S -n 2048
```

## How do I run the loadtests?

The loadtests are in the
[ailoads-loop](https://github.com/mozilla-services/ailoads-loop)
repository.

## Where do I report bugs?

You should report bugs/issues or feature requests via [the loop-server bugzilla
component](https://bugzilla.mozilla.org/enter_bug.cgi?product=Hello (Loop)&component=Server)

## How do I create the release "deploy to stage" bug?

1. Install [deploy-tix](https://github.com/rpappalax/deploy-tix)
2. Run it.

```sh
ticket -r mozilla -a loop-client -e stage -n 0.17.8 -z -u bugzilla_username -p bugzilla_password
```

## Estimate Redis Memory Usage

To estimate Redis usage, checkout the repository and run the `redis_usage.py`
file:

```sh
./redis_usage.py [users] [daily-calls] [monthly-revocation]
```

For instance (for 2M users and 10 calls per day per user)

```sh
$ ./redis_usage.py 2000000 10

loop-server: v0.6.0-DEV
Usage for 2000000 users, with 10 daily calls per user and 0 monthly
revocations is 26569 MBytes
```

The biggest AWS ElastiCache Redis virtual machine is 68GB large so if we want
to handle more that 150M users we will probably want to do some sharding to
have one Redis for calls and another one for user management.

## License

The Loop server code is released under the terms of the
[Mozilla Public License v2.0](http://www.mozilla.org/MPL/2.0/). See the
[`LICENSE`](LICENSE) file at the root of the repository.


CHANGELOG
CHANGELOG
=========

This document describes changes between each past release.

0.22.0 (unreleased)
-------------------

- No changes yet.


0.21.3 (2016-05-26)
-------------------

- Revert sodium dep to 1.0.13 so that Centos 6.5 can install it.
- Revert hiredis dep to 0.2.0 so that Centos 6.5 can install it.
- Revert heapdump dep to 0.3.5 so that Centos 6.5 can install it.


0.21.2 (2016-05-26)
-------------------

- Always use the roomOwnerHmac when upgrading rooms. (Bug 1275652)


0.21.1 (2016-05-17)
-------------------

- Revert sodium dep to 1.0.13 so that Centos 6.5 can install it.
- Revert hiredis dep to 0.2.0 so that Centos 6.5 can install it.


0.21.0 (2016-04-21)
-------------------

- Update rooms sessionId in case tokbox credentials where changed. (#370)
- Update dependencies so that loop-server can run with node 4 (#369)


0.20.1 (2016-03-16)
-------------------

- Revert sodium dep to 1.0.13 so that Centos 6.5 can install it.


0.20.0 (2016-03-03)
-------------------

- Switch from statsd-node to node-statsd to handle tags. (#363)
- Remove loads v1 loadtests. (#364)
- Display pretty exception when tests raises with Mocha. (#365)
- Add a new room action to get metrics about whitelisted shared domains (#365)
- Add a new route to let the client send UX metrics in Google Analytics (#366)
- Upgrade libsodium and add support for node 0.12 (#367)
- Add a __lbhealthcheck__ endpoint (#368)


0.19.3 (2016-02-12)
-------------------

- Add a way to log the loop-client versions in use. (#362)


0.19.2 (2016-01-07)
-------------------

- Update ws module to version 1.0.1 (#360)


0.19.1 (2015-09-18)
-------------------

- Log TokBox errors in heartbeat
- Log Verifier errors in heartbeat
- Fix eslint new configuration errors. (#356)


0.19.0 (2015-09-02)
-------------------

- Logs an error if the Heartbeat fails (#355)
- A user should be able to rejoin with the same session (#353)
- Use the configured push server health endpoint for hearbeat checks (#351)


0.18.1 (2015-08-20)
-------------------

- Fix the status URL for FxA in healthcheck (#350)


0.18.0 (2015-07-29)
-------------------

- Fix 500 when FxA verifier is down now explain why in a 401. (#343)
- Disable the call-url creation endpoint (#337)
- The loadtests now use a FxA test account to place a call (#337)
- Add a check for the Firefox Account verifier in the heartbeat (#345)


0.17.8 (2015-07-24)
-------------------

- Add SimplePush statsd counter (#342)
- Add TokBox statsd counter (#335)
- Log roomConnectionId on action:status calls (#339)
- Handle room creation with null context (#338)


0.17.7 (2015-04-24)
-------------------

- Reject leave on non-joined rooms (#332)
- Dump the heap on uncaught errors (#333)


0.17.6 (2015-04-21)
-------------------

- Add acceptBacklog option to the configuration


0.17.5 (2015-04-06)
-------------------

- Add better error messages to s3.

0.17.4 (2015-04-09)
-------------------

- Downgrade request to 2.45.0, since later versions introduce a latency in our
  requests.


0.17.3 (2015-03-31)
-------------------

- AWS driver did not take bucket configuration into account.
- Add a ``loop.`` prefix on not prefixed keys.


0.17.2 (2015-03-31)
-------------------

- Fix the configuration validator of the filestorage name for AWS.


0.17.1 (2015-03-26)
-------------------

- S3 bucket names can be defined in the configuration.


0.17.0 (2015-03-24)
-------------------

Breaking changes:
~~~~~~~~~~~~~~~~~

- Remove validation on room "status" action (1144851)


New features:
~~~~~~~~~~~~~

- Allow removal of transactions (MULTI) when sharding is enabled (#315)
- Add file system and S3 file storage mechanisms (1141105)


0.16.0 (2015-03-12)
-------------------

Breaking changes:
~~~~~~~~~~~~~~~~~

  - Log the correct number of users in a room (1137882)

New features:
~~~~~~~~~~~~~

  - Handle the status action (1137849)
  - Added a command to count the number of FxA users (1125597)
  - Log tokbox sessionIds associated to the rooms (1134886)
  - Log current room connection id for GET / POSTS on /rooms/{token} (1134884)
  - Log room participant identifier (1134884)
  - Log number of participants for all room actions (1137886)
  - Log the room token on room creation (1137883)
  - Log user type (1140181)

Internal changes:
~~~~~~~~~~~~~~~~~

  - Updated NPM dependencies;
  - Updated linting rules;
  - Loadtests can now precise the SIMPLE_PUSH_URL to use


0.15.6 (2015-02-18)
-------------------

- [bugfix] Update the ping method to not block on existing key.

0.15.5 (2015-02-10)
-------------------

- [tools] Update the migration script to use a redis scan (1125788)
- [bugfix] Fix the heartbeat tokbox handling (1129905)
- [metrics] Integrate with newrelic for performance monitoring (1126889)

0.15.4 (2015-02-02)
-------------------

- [metrics] Add metrics for loop push (1127116)
- [metrics] Time calls to tokbox — bugfix (1125878)

0.15.3 (2015-01-28)
-------------------

- [bugfix] ignore unexisting rooms when deleting multiple rooms
- [bugfix] ignore unexisting rooms when using touchRoomData
- [config] Increase the default HTTP connection pool size (1125777)
- [metrics] Time calls to tokbox (1125878)

0.15.2 (2015-01-23)
-------------------

- [metrics] Do not use loop. prefix for event types (1080133)
- [tools] Add a script to move redis data over (1023396)

0.15.1 (2015-01-19)
-------------------

- [migration] Migration script for room participants
- [feature]   Split USER_UNAVAILABLE from INVALID_PARAMETERS (1120001)

0.15.1 (2015-01-20)
-------------------

- [bugfix] Add a migration script for room participants (1121403)
- [feature] Split out USER_UNAVAILABLE from INVALID_PARAMETERS (1120001)


0.15.0 (2015-01-13)
-------------------

- [bugfix]  Update loadtest with the new room's owner guarented seat. (1111655)
- [bugfix]  Treat user identifier as case-insensitive (1119855)
- [config]  Change logging from winston to mozlog - logs are now with heka

loop/config.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

var convict = require('convict');
var format = require('util').format;
var getHashes = require('crypto').getHashes;
var path = require('path');
var fs = require('fs');

/**
 * Validates the keys are present in the configuration object.
 *
 * @param {List}    keys      A list of keys that must be present.
 * @param {Boolean} options   List of options to use.
 **/
function validateKeys(keys, options) {
  options = options || {};
  var optional = options.optional || false;

  return function(val) {
    if (JSON.stringify(val) === "{}" && optional === true) {
      return;
    }
    if (!optional && !val)
      throw new Error("Should be defined");
    keys.forEach(function(key) {
      if (!val.hasOwnProperty(key))
        throw new Error(format("Should have a %s property", key));
    });
  };
}

/**
 * Build a validator that makes sure of the size and hex format of a key.
 *
 * @param {Integer} size  Number of bytes of the key.
 **/
function hexKeyOfSize(size) {
  return function check(val) {
    if (val === "")
      return;
    if (!new RegExp(format('^[a-fA-FA0-9]{%d}$', size * 2)).test(val)) {
      throw new Error("Should be an " + size +
                      " bytes key encoded as hexadecimal");
    }
  };
}

/**
 * Validates that each channel has an apiKey and apiSecret as well as
 * an optional apiUrl and nothing else. Alse make sure the default
 * channel is defined.
 **/
function tokBoxCredentials(credentials) {
  if (!credentials.hasOwnProperty("default")) {
    throw new Error("Please define the default TokBox channel credentials.");
  }

  var authorizedKeys = ["apiKey", "apiSecret", "apiUrl"];

  function checkKeys(key) {
    if (authorizedKeys.indexOf(key) === -1) {
      throw new Error(key + " configuration value is unknown. " +
                      "Should be one of " + authorizedKeys.join(", ") + ".");
    }
  }

  for (var channel in credentials) {
    // Verify channel keys validity.
    Object.keys(credentials[channel]).forEach(checkKeys);

    if (!credentials[channel].hasOwnProperty("apiKey")) {
      throw new Error(channel + " channel should define an apiKey.");
    }
    if (!credentials[channel].hasOwnProperty("apiSecret")) {
      throw new Error(channel + " channel should define an apiSecret.");
    }
  }
}

var conf = convict({
  env: {
    doc: "The applicaton environment.",
    format: [ "dev", "test", "stage", "prod", "loadtest"],
    default: "dev",
    env: "NODE_ENV"
  },
  ip: {
    doc: "The IP address to bind.",
    format: "ipaddress",
    default: "127.0.0.1",
    env: "IP_ADDRESS"
  },
  port: {
    doc: "The port to bind.",
    format: "port",
    default: 5000,
    env: "PORT"
  },
  acceptBacklog: {
    doc: "The maximum length of the queue of pending connections",
    format: Number,
    default: 511,
    env: "ACCEPT_BACKLOG"
  },
  publicServerAddress: {
    doc: "The public-facing server address",
    format: String,
    default: "localhost:5000",
    env: "SERVER_ADDRESS"
  },
  protocol: {
    doc: "The protocol the server is behind. Should be https behind an ELB.",
    format: String,
    default: "http",
    env: "PROTOCOL"
  },
  macSecret: {
    doc: "The secret for MAC tokens (32 bytes key encoded as hex)",
    format: hexKeyOfSize(32),
    default: "",
    env: "MAC_SECRET"
  },
  encryptionSecret: {
    doc: "The secret for encrypting tokens (16 bytes key encoded as hex)",
    format: hexKeyOfSize(16),
    default: "",
    env: "ENCRYPTING_SECRET"
  },
  userMacSecret: {
    doc: "The secret for hmac-ing userIds (16 bytes key encoded as hex)",
    format: hexKeyOfSize(16),
    default: "",
    env: "USER_MAC_SECRET"
  },
  userMacAlgorithm: {
    doc: "The algorithm that should be used to mac userIds",
    format: function(val) {
      if (getHashes().indexOf(val) === -1) {
        throw new Error("Given hmac algorithm is not supported");
      }
    },
    default: "sha256",
    env: "USER_MAC_ALGORITHM"
  },
  calls: {
    maxSubjectSize: {
      doc: "The maximum number of chars for the subject of a call",
      format: Number,
      default: 124
    }
  },
  callUrls: {
    tokenSize: {
      doc: "The callUrl token size (in bytes).",
      format: Number,
      default: 8
    },
    timeout: {
      doc: "How much time a token is valid for (in hours)",
      format: Number,
      default: 24 * 30 // One month.
    },
    maxTimeout: {
      doc: "The maximum number of hours a token can be valid for.",
      format: Number,
      default: 24 * 30
    },
    webAppUrl: {
      doc: "Loop Web App Home Page.",
      format: "url",
      default: "http://localhost:3000/content/c/{token}",
      env: "WEB_APP_URL"
    }
  },
  displayVersion: {
    doc: "Display the server version on the homepage.",
    default: true,
    format: Boolean
  },
  storage: {
    engine: {
      doc: "engine type",
      format: String,
      default: "redis"
    },
    settings: {
      doc: "js object of options to pass to the storage engine",
      format: Object,
      default: {
        port: 6379,
        host: 'localhost'
      }
    }
  },
  filestorage: {
    engine: {
      doc: "engine type",
      format: ["filesystem", "aws"],
      default: "filesystem"
    },
    settings: {
      doc: "js object of options to pass to the files engine",
      format: Object,
      default: {
        base_dir: "/tmp"
      }
    }
  },
  pubsub: {
    doc: "js object of options to pass to the pubsub engine",
    format: Object,
    default: {
      port: 6379,
      host: 'localhost',
      skipConfig: false
    }
  },
  fakeTokBox: {
    doc: "Mock TokBox calls",
    format: Boolean,
    default: false
  },
  fakeTokBoxURL: {
    doc: "URL where to Mock TokBox calls",
    format: String,
    default: "https://call.stage.mozaws.net/"
  },
  tokBox: {
    apiUrl: {
      doc: 'api endpoint for tokbox',
      format: String,
      default: "https://api.opentok.com"
    },
    credentials: {
      doc: "api credentials based on a channel.",
      format: tokBoxCredentials,
      default: {}
    },
    tokenDuration: {
      doc: 'how long api tokens are valid for in seconds',
      format: "nat",
      default: 24 * 3600
    },
    retryOnError: {
      doc: 'how many times to retry on error',
      format: "nat",
      default: 3
    },
    timeout: {
      doc: "Timeout for requests when trying to create the session (ms)",
      format: Number,
      default: 2000
    }
  },
  sentryDSN: {
    doc: "Sentry DSN",
    format: function(val) {
      if (!(typeof val === "string" || val === false)) {
        throw new Error("should be either a sentryDSN or 'false'");
      }
    },
    default: false,
    env: "SENTRY_DSN"
  },
  statsd: {
    doc: "Statsd configuration",
    format: validateKeys(['port', 'host'], {'optional': true}),
    default: {}
  },
  statsdEnabled: {
    doc: "Defines if statsd is enabled or not",
    format: Boolean,
    default: false
  },
  allowedOrigins: {
    doc: "Authorized origins for cross-origin requests.",
    format: Array,
    default: ['http://localhost:3000']
  },
  retryAfter: {
    doc: "Seconds to wait for on 503",
    format: Number,
    default: 30
  },
  fxaAudiences: {
    doc: "List of accepted fxa audiences.",
    format: Array,
    default: []
  },
  fxaVerifier: {
    doc: "The Firefox Accounts verifier url",
    format: String,
    env: "FXA_VERIFIER",
    default: "https://verifier.accounts.firefox.com/v2"
  },
  fxaTrustedIssuers: {
    doc: "The list of Firefox Accounts trusted issuers",
    format: Array,
    default: ["api.accounts.firefox.com"]
  },
  hawkIdSecret: {
    doc: "The secret for hmac-ing the hawk id (16 bytes key encoded as hex)",
    format: hexKeyOfSize(16),
    default: "",
    env: "HAWK_ID_SECRET"
  },
  hawkSessionDuration: {
    doc: "The duration of hawk credentials (in seconds)",
    format: Number,
    default: 3600 * 24 * 30 // One month.
  },
  callDuration: {
    doc: "The duration we want to store the call info (in seconds)",
    format: Number,
    default: 60
  },
  maxHTTPSockets: {
    doc: "The maximum of HTTP sockets to use when doing requests",
    format: Number,
    default: 10000000
  },
  heartbeatTimeout: {
    doc: "Timeout for requests when doing heartbeat checks (ms)",
    format: Number,
    default: 2000
  },
  timers: {
    supervisoryDuration: {
      doc: "Websocket timeout for the supervisory timer (seconds)",
      format: Number,
      default: 10
    },
    ringingDuration: {
      doc: "Websocket timeout for the ringing timer (seconds)",
      format: Number,
      default: 30
    },
    connectionDuration: {
      doc: "Websocket timeout for the connection timer (seconds)",
      format: Number,
      default: 10
    }
  },
  progressURLEndpoint: {
    doc: "The endpoint to use for the progressURL.",
    format: String,
    default: "/websocket"
  },
  i18n: {
    defaultLang: {
      format: String,
      default: 'en-US'
    }
  },
  pushServerURIs: {
    doc: "An array of push server URIs",
    format: Array,
    default: ["wss://push.services.mozilla.com/"]
  },
  fxaOAuth: {
    activated: {
      doc: "Set to false if you want to deactivate FxA-OAuth on this instance.",
      format: Boolean,
      default: true
    },
    client_id: {
      doc: "The FxA client_id (8 bytes key encoded as hex)",
      format: hexKeyOfSize(8),
      default: ""
    },
    client_secret: {
      doc: "The FxA client secret (32 bytes key encoded as hex)",
      format: hexKeyOfSize(32),
      default: ""
    },
    oauth_uri: {
      doc: "The location of the FxA OAuth server.",
      format: "url",
      default: "https://oauth.accounts.firefox.com/v1"
    },
    content_uri: {
      doc: "The location of the FxA content server.",
      format: "url",
      default: "https://accounts.firefox.com"
    },
    redirect_uri: {
      doc: "The redirect_uri.",
      format: String,
      default: "urn:ietf:wg:oauth:2.0:fx:webchannel"
    },
    profile_uri: {
      doc: "The FxA profile uri.",
      format: "url",
      default: "https://profile.firefox.com/v1"
    },
    scope: {
      doc: "The scope we're requesting access to",
      format: String,
      default: "profile"
    }
  },
  logRequests: {
    activated: {
      doc: "Defines if requests should be logged to Stdout",
      default: false,
      format: Boolean
    },
    consoleDateFormat: {
      doc: "Date format of the logging line.",
      format: String,
      default: "%y/%b/%d %H:%M:%S"
    }
  },
  hekaMetrics: {
    activated: {
      doc: "Defines if metrics should be directed to hekad",
      default: false,
      format: Boolean
    },
    debug: {
      doc: "Heka logger display development logs",
      format: Boolean,
      default: false
    },
    level: {
      doc: "Log level.",
      format: String,
      default: "INFO"
    },
    fmt: {
      doc: "Log level.",
      format: String,
      default: "heka"
    }
  },
  newRelic: {
    activated: {
      doc: "Defines if newrelic is activated or not",
      default: false,
      format: Boolean
    },
    licenceKey: {
      doc: "New Relic licence key",
      format: String,
      default: ""
    },
    loggingLevel: {
      doc: "Logging level to use",
      format: String,
      default: "info"
    },
    appName: {
      doc: "New Relic application name",
      format: String,
      default: "Loop Server"
    }
  },
  dumpHeap: {
    activated: {
      doc: "Defines if uncaught exceptions should dump snapshots",
      default: true,
      format: Boolean
    },
    location: {
      doc: "Location where to output the files (directory).",
      format: String,
      default: '/tmp'
    }
  },
  rooms: {
    defaultTTL: {
      doc: "The default TTL for a room (in hours)",
      format: Number,
      default: 8 * 7 * 24 // 8 weeks
    },
    maxTTL: {
      doc: "The maximum TTL for a room (in hours) allowed by the server",
      format: Number,
      default: 8 * 7 * 24 // 8 weeks
    },
    extendTTL: {
      doc: "The new TTL for a room (in hours) after a participant join.",
      format: Number,
      default: 8 * 7 * 24 // 8 weeks
    },
    participantTTL: {
      doc: "The TTL (in seconds) for a participant in the room",
      format: Number,
      default: 5 * 60  // 5 minutes
    },
    deletedTTL: {
      doc: "The TTL (in seconds) for a room delete notification to be saved",
      format: Number,
      default: 27 * 60  // 27 minutes
    },
    maxSize: {
      doc: "The maximum size of a room",
      format: Number,
      default: 5
    },
    maxRoomNameSize: {
      doc: "The maximum number of chars to name a room",
      format: Number,
      default: 100
    },
    maxRoomOwnerSize: {
      doc: "The maximum number of chars for the owner of a room",
      format: Number,
      default: 255
    },
    tokenSize: {
      doc: "The room token size (in bytes).",
      format: Number,
      default: 8
    },
    webAppUrl: {
      doc: "Loop Web App rooms url.",
      format: "url",
      default: "http://localhost:3000/content/{token}",
      env: "ROOMS_WEB_APP_URL"
    },
    HKDFSalt: {
      doc: "The salt that will be used to cipher profile data " +
           "(16 bytes key encoded as hex)",
      format: hexKeyOfSize(16),
      default: "",
      env: "ROOMS_HKDF_SECRET"
    }
  }
});


// handle configuration files.  you can specify a CSV list of configuration
// files to process, which will be overlayed in order, in the CONFIG_FILES
// environment variable. By default, the ../config/<env>.json file is loaded.

var envConfig = path.join(__dirname, '/../config', conf.get('env') + '.json');
var files = (envConfig + ',' + process.env.CONFIG_FILES)
    .split(',')
    .filter(fs.existsSync);

conf.loadFile(files);
conf.validate();

if (conf.get('macSecret') === "")
  throw "Please define macSecret in your configuration file";

if (conf.get('rooms').HKDFSalt === "")
    throw "Please define rooms.HKDFSalt in your configuration file";

if (conf.get('encryptionSecret') === "")
  throw "Please define encryptionSecret in your configuration file";

if (conf.get('allowedOrigins') === "") {
  throw "Please define the list of allowed origins for CORS.";
}

if (conf.get('fxaAudiences').length === 0) {
  throw "Please define the list of allowed Firefox Accounts audiences";
}

if (conf.get('hawkSessionDuration') <
    conf.get('callUrls').maxTimeout * 60 * 60) {
  throw "hawkSessionDuration should be longer or equal to callUrls.maxTimeout";
}

if (conf.get('fxaOAuth').activated && conf.get('fxaOAuth').client_id === "") {
  throw "fxaOAuth is activated but not well configured. " +
    "Set fxaOAuth.activated config key to false to continue";
}

// Configure S3 the environment variable for tests.
if (conf.get('env') === 'test') {
  process.env.AWS_ACCESS_KEY_ID = 'TESTME';
  process.env.AWS_SECRET_ACCESS_KEY = 'TESTME';
}

// Verify timers
var timers = conf.get("timers");
var minCallDuration = timers.supervisoryDuration +
  timers.ringingDuration +
  timers.connectionDuration;
if (minCallDuration > conf.get("callDuration")) {
  throw "the callDuration should be at least " + minCallDuration + " seconds";
}

module.exports = {
  conf: conf,
  hexKeyOfSize: hexKeyOfSize,
  validateKeys: validateKeys
};
